Accept arrays in parse_config's method_missing

rgeissert opened the pull request:
## Description
A single-element array allows Rspec's its behaviour to be worked around
and allow options containing dots to be tested using its.

This is already implemented by resources such as `json` and those based
on it (e.g. xml).

N.b. I feel like I'm missing something with regards to the documentation as I see it is in two directories. Is there some automagic tool I forgot to run?
Note as well that in the documentation I mention version `4.23.9` as the future version including this change, so feel free to correct it as needed.

## Related Issue
Related to issue inspec#875.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
- [x] I have read the **CONTRIBUTING** document.

rgeissert committed:
Accept arrays in parse_config's method_missing

A single-element array allows Rspec's its behaviour to be worked around
and allow options containing dots to be tested using its.

This is already implemented by resources such as `json` and those based
on it (e.g. xml).

Related to issue #875.

Signed-off-by: Raphael Geissert <atomo64@gmail.com>

rgeissert committed:
Move the options documentation under the correct section

Obvious fix, the options are not related to the matchers.

rgeissert committed:
Add missing options section to the parse_config documentation

Obvious fix, the section was missing. Complete it with an example
from parse_config_file.

james-stocks commented on test/unit/resources/parse_config_test.rb:
```diff
@@ -24,4 +24,13 @@
     _(resource.send("kernel.domainname")).must_equal "example.com"
   end
+
+  it "parse_config resource accepts arrays" do
```
Maybe we should mention the rspec `its` case here. In future, we might read this test and wonder why we needed to support this case.

rgeissert replied:
I modified it to say `parse_config resource accepts arrays due to rspec's its behavior`, does that look good enough to you?

james-stocks replied:
Brilliant, thank you!

james-stocks reviewed (approved):
This is great, thank you @rgeissert !
I have added our docs team as reviewer.

james-stocks commented on docs-chef-io/content/inspec/resources/parse_config.md:
```diff
@@ -53,9 +53,20 @@ where each test
+        multiple_values: true
+      }
+
+    output = command('some-command').stdout
```
We should use this new example code; but bear in mind that we increasingly see issues with using Ruby variable assignment like this, because the code executes unconditionally even if the control is skipped/waived/excluded.

The only current solution is to write a custom resource that wraps up the code. In future we want InSpec to to properly skip code for excluded controls...

clintoncwolfe commented on lib/inspec/resources/parse_config.rb:
```diff
@@ -55,8 +55,11 @@ def initialize(content = nil, useropts = nil)
-      read_params[name.to_s]
+    def method_missing(*name)
+      # catch bahavior of rspec its implementation
+      # @see https://github.com/rspec/rspec-its/blob/master/lib/rspec/its.rb#L110
```
Unfortunately, this URL is already bad - consider replacing with a permalink

rgeissert replied:
Fixed by making it point to tag v1.2.0, the version of rspec-its currently in use.

clintoncwolfe commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+### Test a configuration setting containing dots
+
+    describe parse_config_file('/etc/sysctl.conf') do
+     its(['kernel.domainname']) { should eq 'example.com' }
```
This is just opinion.

This just feels strange - I think this may be a case of implementation driving UX. If I were a user, and I wanted to do a query containing a dot, even assuming I knew it had to be escaped, the typical thing would be to use a slash, not to enclose the query in brackets.

I agree the need is valid, and here we wish we could hijack `rspec-its` dot-processing facility entirely. So why not do so? In method_missing, why not explicitly search for dots and transform the call as needed to fool `rspec-its`? Then there may be no need to alter the DSL syntax at all - you may be able to have:

```
its('kernel.domainname') { should eq 'example.com' }
```

simply work.

rgeissert replied:
From my understanding of rspec-its' code our method_missing won't ever see the dot due to the code at:
https://github.com/rspec/rspec-its/blob/v1.2.0/lib/rspec/its.rb#L113

I drafted some code to return some magic object from method_missing so that the calls to `send` Just Work.
It fails, however, whenever the hash contains "foo.bar" and "foo.bar.baz" and one wants to get the latter.

If that sounds acceptable I can use that instead of accepting arrays (like other resources do already).

I don't know how it should behave with regards to_sth methods. E.g. only "foo.bar.baz" exists, but the caller wants 'foo.bar' only.
On the plus side, `its('foo.bar.baz.downcase')` just works.

clintoncwolfe replied:
Ah, thank you, that code reference cleared up the situation for me. I think your existing approach using the array syntax is an acceptable compromise.

rgeissert replied:
If it's okay for you to leave it as-is then I think this pull request is ready from my POV.

rgeissert committed:
Fix the reference to rspec-its' behaviour with arrays

Make it point to the version currently in use.

Obvious fix, the link to master is useless as the code moved.

mjingle commented on docs-chef-io/content/inspec/resources/parse_config.md:
```diff
@@ -114,3 +125,12 @@ Use `standalone_comments: false`, to parse the following:
+
+## Examples
+
+This resource is based on the `parse_config_file` resource. As such please refer to the examples of the [`parse_config_file`](/inspec/resources/parse_config_file) resource.
```
```suggestion
This resource is based on the `parse_config_file` resource. See the [`parse_config_file`](/inspec/resources/parse_config_file) resource for examples.
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config.md:
```diff
@@ -53,9 +53,20 @@ where each test
+## Options
 
-For a full list of available matchers, please visit our [matchers page](/inspec/matchers/).
+This resource supports the following options for parsing configuration data. Use them in an `options` block stated outside of (and immediately before) the actual test:
```
```suggestion
This resource supports multiple options to parse configuration data. Use the options in an `options` block stated outside of (and immediately before) the actual test. For example:
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+     its('main') { should include('gpgcheck' => '1') }
+    end
+
+### Test a configuration setting containing dots
```
@rgeissert When referring to "dots," are you talking about periods (`.`) within a URL, a specific technical file, or something else?

rgeissert replied:
@mjingle It should read: "Test a configuration setting whose name contains dots" (or periods, or full stops, as you prefer :)). As in: `author.name = Raphael` in a configuration file.

mjingle replied:
@rgeissert Awesome! Thank you for confirming. That will help inform my suggested edits to you, and then we'll get this docs part finished.

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -50,6 +50,13 @@ where each test
 - May run a command to `stdout`, and then run the test against that output
 - May use options to define how configuration data is to be parsed
 
+### Option names containing dots
```
```suggestion
### Option Names Containing Periods
```

Website formatting prefers to have the first letter of each word in a header capitalized.

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -50,6 +50,13 @@ where each test
 
+### Option names containing dots
+
+A possibly unexpected behaviour may occur when attempting to access option names containing dots with `its()`. A known behaviour, there are two ways to work around it:
```
```suggestion
A possible behavior may occur when attempting to access option names containing periods with `its()`. There are two ways to work around it:
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+
+The following examples show how to use this Chef InSpec audit resource.
+
+### Test a configuration setting
```
```suggestion
### Test A Configuration Setting
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+     its('PARAM_X') { should eq 'Y' }
+    end
+
+### Use options, and then test a configuration setting
```
```suggestion
### Use Options And Then Test A Configuration Setting
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+     its('PARAM_X') { should include 'Y' }
+    end
+
+### Test a file with an ini-like structure (such as a yum.conf)
```
```suggestion
### Test A File With An INI File Structure
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+    end
+
+### Test a file with an ini-like structure (such as a yum.conf)
+
```
```suggestion
`yum.conf` is one example of an INI file structure type. 

```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -145,3 +126,35 @@ Use `standalone_comments: false`, to parse the following:
+     its('main') { should include('gpgcheck' => '1') }
+    end
+
+### Test a configuration setting containing dots
```
```suggestion
### Test A Configuration Setting Containing Periods
```

mjingle commented on docs-chef-io/content/inspec/resources/parse_config_file.md:
```diff
@@ -50,6 +50,13 @@ where each test
+A possibly unexpected behaviour may occur when attempting to access option names containing dots with `its()`. A known behaviour, there are two ways to work around it:
+
+* Access the option by using the `params` attribute of the returned resource object
+* Since 4.23.9, `its` can be used by passing the option name in a single-element array (cf. examples)
```
```suggestion
* Since 4.23.9, `its` can be used by passing the option name in a single-element array. See the `parse config file` examples.
```

mjingle replied:
After this PR merges, I will insert an internal reference link to the `parse config file` Examples section as a part of a link clean-up project in the docs. So know that this is planned and won't be omitted in the long term.

rgeissert replied:
I bumped the version number but it should be fixed once the PR is actually merged.

mjingle reviewed (commented):
Okay, @rgeissert! I've suggested all of my suggested changes in the docs portion. After these are fixed, you should be good from a docs perspective. Thank you for your patience! :)

Commit:
Apply suggestions from the doc review

Signed-off-by: Raphael Geissert <atomo64@gmail.com>

mjingle reviewed (approved):
Good to go from a docs perspective. Thank you again!

clintoncwolfe reviewed (approved):
Thanks!
